File: Part 2/A. ABC String.cpp
//Solver : Nahin the Coder
//Date : 28-07-2021

#include <bits/stdc++.h>
using namespace std;

int main() 
{
	int test;
	cin>>test;
	
	while(test--)
	{
		string str;
		cin>>str;
		int n = str.length();
		if(str[0] == str[n-1])
		{
			cout<<"NO\n";
			continue;
		}
		int closing = 0;
		int opening = 0;
	    for(int i=0; i<n; i++)
	    {
	    	if(str[0] == str[i]) opening++;
	    	else opening--;
	    	//break
			if(opening<0) break;
			
		}
		for(int i=0; i<n; i++)
	    {
			if(str[i] == str[n-1]) closing--;
			else closing++; 
			if(closing < 0) break;
		}
		if(opening == 0 or closing == 0)
		{
			cout<<"YES\n";
		}
		else
		{
			cout<<"NO\n";
		}
	}
	
	
	
	
    return 0;
}

File: Solved by Python/A. ABC String.py
# Nahin the Coder
# Date : 28 - 07 - 2021


testcase = int(input())
while testcase > 0:
    st = input()
    n = len(st)
    if st[0] == st[n-1]:
        print("NO")
    else:
        opening = int(0)
        closing = int(0)
        for i in range(n):
            if st[0] == st[i]:
                opening = opening + 1
            else:
                opening = opening - 1
            if opening < 0:
                break
        for i in range(n):
            if st[n-1] == st[i]:
                closing = closing - 1
            else:
                closing = closing + 1
            if closing < 0:
                break
        if closing == 0 or opening == 0:
            print("YES")
        else:
            print("NO")
    testcase = testcase - 1
    
